### PR TITLE
Add persona views and templates

### PR DESCRIPTION
**Pull request by codiebeulaine**

### CODE REVIEW

**Comment by codiebeulaine** [comment]
@miltontony @saeedmb8 read for review. tests and pretty template to follow in separate PR

File: molo/usermetadata/views.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+def SetPersonaView(request, persona_id):
+    persona = get_object_or_404(PersonaPage, pk=persona_id)
+    if not ('MOLO_PERSONA_SELECTION') in request.session:
+        request.session['MOLO_PERSONA_SELECTION'] = persona.slug
+        return HttpResponseRedirect(request.GET.get('next', '/'))
+
+
+def SkipPersonaView(request):
+    request.session['MOLO_PERSONA_SELECTION'] = ''
```
Comment (smarzban) [inline molo/usermetadata/views.py]: @codiebeulaine lets store `'skip'` instead of  `' '`

File: molo/usermetadata/urls.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+
+urlpatterns = patterns(
+    '',
+    url(r'^persona/$', views.PersonaView.as_view(), name='persona'),
+    url(
+        r'^persona/(?P<persona_id>\d+)$',
+        views.SetPersonaView, name='set_persona'),
+    url(
+        r'^persona/skip$',
```
Comment (smarzban) [inline molo/usermetadata/urls.py]: @codiebeulaine this should be `r'^persona/skip/$',`

**Comment by codiebeulaine** [comment]
@saeedmb8 ready for review again :)

File: molo/usermetadata/views.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class PersonaView(TemplateView):
+    template_name = 'persona.html'
+
+    def get_context_data(self, *args, **kwargs):
+        persona_pages = PersonaPage.objects.live().all()
+        context = super(PersonaView, self).get_context_data(*args, **kwargs)
+        context.update({
+            'persona_pages': persona_pages,
```
Comment (smarzban) [inline molo/usermetadata/views.py]: @codiebeulaine here we need to check see if a persona page has translation or not, so you need to do `'persona_pages': [a.get_translation_for(locale_code) or a for a in persona_pages]`

File: molo/usermetadata/views.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+from django.http import HttpResponseRedirect
+
+from molo.usermetadata.models import PersonaPage
+
+
+class PersonaView(TemplateView):
+    template_name = 'persona.html'
+
+    def get_context_data(self, *args, **kwargs):
+        persona_pages = PersonaPage.objects.live().all()
```
Comment (smarzban) [inline molo/usermetadata/views.py]: @codiebeulaine and you need to get the locale code:

```
context = copy(context)
locale_code = context.get('locale_code')
```

File: molo/usermetadata/views.py (hunk 3 above)
Comment (smarzban) [inline molo/usermetadata/views.py]: @codiebeulaine please also add tests for the translated persona pages as well :)

File: molo/usermetadata/views.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        context = super(PersonaView, self).get_context_data(*args, **kwargs)
+        locale_code = context.get('locale_code')
+        context.update({
+            'persona_pages':
+            [a.get_translation_for(locale_code) or a for a in persona_pages],
```
Comment (justinvdm) [inline molo/usermetadata/views.py:19]: When we add tests, can we add a test to check that we use translated pages if a language with `locale_code` exists, and use the original page if no language with `locale_code` exists?

**Comment by justinvdm** [comment]
Other than the request to add tests, looks good :)

File: molo/usermetadata/views.py (hunk 5 above)
Comment (codiebeulaine) [inline molo/usermetadata/views.py:19]: sure :)

File: molo/usermetadata/middleware.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+            'django-admin/',
+            '/locale/'
+        ]
+        if any([path for path in exclude if request.path.startswith(path)]):
+            return None
+
+        if not ('MOLO_PERSONA_SELECTION') in request.session:
+            url = reverse('molo.usermetadata:persona')
+            url = '%s?next=%s' % (url, request.path)
+            return HttpResponseRedirect(url)
```
Comment (miltontony) [inline molo/usermetadata/middleware.py]: [use redirect shortcut](https://docs.djangoproject.com/en/1.9/topics/http/shortcuts/#redirect)

File: molo/usermetadata/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    def test_skip_persona_page(self):
+
+        response = self.client.get('/')
+        self.assertRedirects(
+            response, reverse('molo.usermetadata:persona') + '?next=/')
+
+        response = self.client.get('%s?next=%s' % ((
+            reverse('molo.usermetadata:skip_persona')), '/'))
+        self.assertRedirects(response, '/')
```
Comment (miltontony) [inline molo/usermetadata/tests/test_views.py:103]: Add an assert to make sure that subsequent requests don't redirect (i.e cookie is set)

**Comment by miltontony** [comment]
:+1:

**Comment by smarzban** [comment]
@miltontony @codiebeulaine ready for review :)

File: molo/usermetadata/middleware.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            exclude = [
+                settings.MEDIA_URL,
+                settings.STATIC_URL,
+                reverse('molo.usermetadata:persona'),
+                reverse('health'),
+                reverse('versions'),
+                '/admin/',
+                'django-admin/',
+                '/import/',
+                '/import/',
```
Comment (miltontony) [inline molo/usermetadata/middleware.py]: @saeedmb8 duplicate

**Comment by smarzban** [comment]
@codiebeulaine @miltontony are you happy to I land this PR?

**Comment by codiebeulaine** [comment]
@saeedmb8 👍

File: molo/usermetadata/middleware.py [hunk 9]
```diff
@@ -1,31 +1,38 @@
...
-            '/import/',
-            '/locale/'
-        ]
 
-        if hasattr(settings, 'PERSONA_IGNORE_PATH'):
-            exclude += settings.PERSONA_IGNORE_PATH
+        site = Site.objects.get(is_default_site=True)
+        setting = SettingsProxy(site)
+        persona_settings = setting['usermetadata']['PersonaeSettings']
+        if persona_settings.persona_required:
```
Comment (miltontony) [inline molo/usermetadata/middleware.py]: @saeedmb8 I think you should just do

``` python
if not persona_settings.persona_required:
    return None
```

**Comment by miltontony** [comment]
:+1: :cake: :cake: :cake: :cake: :cake: :cake:

**Pull request merged**